### PR TITLE
feat: add file to compute actions and get action mask for N5 neighbor…

### PR DESCRIPTION
**Pull request by MarcBoelle3**
…hood

get_critical_operations function: creates an array containing, for each operation, information helpful to determine valid actions.
get_action_mask_n5: outputs the n5 action mask based on critical block information
select_operations_to_switch: from an action (chosen operation, left or right), outputs the according pair of operations to switch.

### CODE REVIEW

File: jumanji/environments/packing/job_shop/improvement/get_actions.py [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+    )
+    # All ops are initially marked as right ends (updated during scan)
+    critical_block_info = critical_block_info.at[:, CBFields.IS_RIGHT].set(1)
+    # Initialize neighbors and right_end to -1
+    critical_block_info = critical_block_info.at[:, CBFields.LEFT_NEIGHBOR].set(-1)
+    critical_block_info = critical_block_info.at[:, CBFields.RIGHT_NEIGHBOR].set(-1)
+    critical_block_info = critical_block_info.at[:, CBFields.RIGHT_END].set(-1)
+
+    # Initialize right_end_array to -1
+    # Right end array is used to
```
Comment (mintiti) [inline jumanji/environments/packing/job_shop/improvement/get_actions.py]: Incomplete comment i think :)

File: jumanji/environments/packing/job_shop/improvement/get_actions.py [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+    def cond_fun(loop_state: Tuple[jnp.int32, Tuple[chex.Array, chex.Array]]) -> jnp.bool_:
+        """Condition function for the while loop.
+        The iteration stops when we have processed all the valid critical block pairs (i.e.
+        while start_idx and end_idx are both not -1).
+        """
+        i, _ = loop_state
+        start_idx, end_idx = critical_block_pairs[i]
+        return jnp.logical_and(i < num_ops_total, jnp.logical_and(start_idx != -1, end_idx != -1))
+
+    def body_fun(
```
Comment (mintiti) [inline jumanji/environments/packing/job_shop/improvement/get_actions.py]: I don't know if this comment should be on this function or the while loop, but can you quickly explain the core of the algorithm you came up with here ? Mainly what you explained orally, how the data flows in the while loop

File: jumanji/environments/packing/job_shop/improvement/get_actions.py [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+    # with index equal to num_critical_block.
+    final_critical_block_info = final_critical_block_info.at[:, CBFields.IS_LEFT].set(
+        final_critical_block_info[:, CBFields.BLOCK_ID] == jnp.arange(num_ops_total)
+    )
+
+    return final_critical_block_info
+
+
+def get_action_mask_n5(critical_block_info: chex.Array, max_num_ops: int) -> chex.Array:
+    """Get the mask of valid actions for the N5 neighborhood.
```
Comment (mintiti) [inline jumanji/environments/packing/job_shop/improvement/get_actions.py:230]: Do we explain somewhere what N5/N6 means ? might be good to explain it somewhat or put a reference to some external document somewhere

File: jumanji/environments/packing/job_shop/improvement/get_actions_test.py [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+                    [False, False],
+                    [True, False],
+                    [False, False],
+                    [False, False],
+                    [False, False],
+                ]
+            )
+        )
+
+    def test_get_action_mask_n5_jit(
```
Comment (mintiti) [inline jumanji/environments/packing/job_shop/improvement/get_actions_test.py]: Can we add a few more tests for N5 ?
I think there might be edge cases we havnt tested ? This is basic tests with 1 example. 
Maybe we can find a few edge cases that might be problematic ?

Also i would probably separate the test suites between action mask, getting critical block info and N6/N5 selection ?

Each of them in general could be tested a bit more robustly maybe ?

**Review by mintiti — CHANGES_REQUESTED** [review]
Really good code, especially given the complexity of it. 
Well documented and all
I would just request better tests for the critical parts of the algorithm. 
We already have basic tests that the algorithm works, can we think about edge cases ? What other prooperties should the action mask in a N5 case have theoretically ? Can we test it ?
